content/posts/how-to-evolve-application-add-quality/index.md

```diff
@@ -0,0 +1,139 @@
+}
+
+````
+* Add configuration 
+````groovy
+sonar {
+    properties {
+        property "sonar.projectkey", "happraisal"
+        property "sonar.projectName", "happraisal"
+        property "sonar.host.url", "http://localhost:9000"
+        property "sonar.coverage.jacoco.xmlReportPaths", "build/reports/jacoco/all-tests/jacocoAllTestReport.xml"
+        property "sonar.sources", "src/main,build.gradle,gradle"
+        property "sonar.tests", "src/test,src/integration-test,src/system-test"
+        property "sonar.junit.reportPaths", "build/test-results/test,build/test-results/integrationTest,build/test-results/systemTest"
+    }
+}
+````
+
+#### Export results
+
+From intelliJ, via configuration or terminal you can execute :
+```` 
+clean build jacocoTestReport sonar -Dsonar.login=<<YOUR_ID>> -Dsonar.password=<<YOU_PASSWORD>>
+
+````
+
+If you have created a token you can run 
+
+
+````
+./gradlew clean build jacocoTestReport sonar -Dsonar.login=<<TOUR_TOKEN>>
+
+````
+
+### Multi module
+In project with multi-module each module must execute sonar task.
+
+
+## Tools around Sonar
+### Sonar API
+
+Sonar provides some apis to manage project, quality profile, quality gate and others. 
+
+Details can be find [here](https://next.sonarqube.com/sonarqube/web_api/)
+You can use tools like postman or curl to use these apis, but the first thing that you must do is to create a token.
+
+To create a token, after you login  go to your profil and in the security tab and create a token.
+
+For some apis you must create a token with administrator authorizations like 
+* /api/qualitygates/select
+* /api/projects/search
+
+
+
+### Sonarlint
+Sonarlint is a tool that can be used to analyze your code. 
+
+#### Configuration 
+* Add plugin sonarlint in IntelliJ
+* Go to Sonarlint settings in Tools/Sonarlint menu
+ ![sonarlint-settings.png](sonarlint-settings.png)
+
+* Add  a SonarCloud or Sonarqube connection.  We select Sonarcloud
+![sonarlint-create-connexion.png](sonarlint-create-connexion.png)
+
+* Click on the link to connect to Sonarcloud service
+
+![sonarlint-sonarcloud-login.png](sonarlint-sonarcloud-login.png)
+
+* Set token or create one from the link
+![sonarlint-sonarcloud-connexion-token.png](sonarlint-sonarcloud-connexion-token.png)
+
+* Select your organization
+![sonarlint-sonarcloud-organization-choice.png](sonarlint-sonarcloud-organization-choice.png)
+
+* Select notification configuration
+![sonarlint-sonarcloud-configuration-notification.png](sonarlint-sonarcloud-configuration-notification.png)
+
+* Approve the creation 
+![sonarlint-sonarcloud-success.png](sonarlint-sonarcloud-success.png)
+* bind the project with the connexion
+![sonarlint-bind-project-with-connexion.png](sonarlint-bind-project-with-connexion.png)
+
+#### use
+In intelliJ there is a sonarlint window
+![intellij-sonarlint-window.png](intellij-sonarlint-window.png)
+
+In this window: 
+
+* form current file tab you can analyze the file selected and displayed in IntelliJ
```

Comment:
```suggestion
* form the _Current File_ tab you can analyze the file selected and displayed in IntelliJ
```

Reply:
@dcremonini  Ok but why 'the _Current File_' ?

Reply:
Because there a rule about using the "the" article that says: "Use the article _the_ when an adjective, phrase, or clause describing the noun clarifies or restricts its identity."

In this case there are some tabs and _Current File_ identifies a specific tab.
See [here] (https://www.butte.edu/departments/cas/tipsheets/grammar/articles.html#:~:text=Use%20the%20article%20the%20when%20an%20adjective%2C%20phrase%2C%20or%20clause%20describing%20the%20noun%20clarifies%20or%20restricts%20its%20identity.).

Reply:
Done